File: src/sequential_text_parser.js
const logger = require('logging');
const {
  EndTokenNotFoundParserException,
  InvalidTagParserException,
  StartTokenNotFoundParserException,
  TagNotFoundParserException
} = require('./exceptions');

const TextParser = require('./text_parser');

class SequentialTextParser extends TextParser {
  constructor() {
    super();
  }

  parse(tag_name, replace_map = {}) {
    for (const parameter of this.parameters) {
      if (parameter.name === tag_name) {
        let start_tag = parameter.start_tag.toUpperCase();
        let end_tag = parameter.end_tag.toUpperCase();

        for (const [old, value] of Object.entries(replace_map)) {
          if (old && value !== undefined) {
            start_tag = start_tag.replace(old.toUpperCase(), value.toUpperCase());
            end_tag = end_tag.replace(old.toUpperCase(), value.toUpperCase());
          }
        }

        if (!start_tag) {
          console.log(`Error while processing tag: ${tag_name}.`);
          throw new InvalidTagParserException(tag_name);
        }

        let start_block, token, end_block;
        [start_block, token, end_block] = this.partition(this.cursor_upper, start_tag);

        if (!token) {
          console.log(`Start token not found for tag: ${tag_name}.`);
          throw new StartTokenNotFoundParserException(tag_name);
        }

        const offset = start_block.length + token.length;

        if (end_tag) {
          [start_block, token, end_block] = this.partition(end_block, end_tag);

          if (!token) {
            console.log(`End token not found for tag: ${tag_name}.`);
            throw new EndTokenNotFoundParserException(tag_name);
          }

          this.field_upper = start_block;
          this.cursor_upper = end_block;
        } else {
          [this.field_upper, token, this.cursor_upper] = this.partition(end_block, start_tag);
          this.cursor_upper = token + this.cursor_upper;
        }

        this.field = this.cursor.substring(offset, offset + this.field_upper.length);
        this.cursor = this.cursor.substring(this.cursor.length - this.cursor_upper.length);

        return this;
      }
    }

    console.log(`Tag ${tag_name} not found during parsing.`);
    throw new TagNotFoundParserException(tag_name);
  }
}

module.exports = {SequentialTextParser};

File: src/text_parser/sequential_text_parser.py
import logging

from exceptions.parser_exceptions import (
    EndTokenNotFoundParserException,
    InvalidTagParserException,
    StartTokenNotFoundParserException,
    TagNotFoundParserException,
)
from text_parser.text_parser import TextParser

logger = logging.getLogger(__name__)


class SequentialTextParser(TextParser):
    """A text parser that sequentially searches for tags and
    extracts content between start and end tokens."""

    def parse(self, tag_name: str, replace_map={}):
        for parameter in self.parameters:
            if parameter.name == tag_name:
                start_tag = parameter.start_tag.upper()
                end_tag = parameter.end_tag.upper()

                for old, new in replace_map.items():
                    if old and (new is not None):
                        start_tag = start_tag.replace(old.upper(), new.upper())
                        end_tag = end_tag.replace(old.upper(), new.upper())

                if not start_tag:
                    logger.warning(f"Error while processing tag: '{tag_name}'.")
                    raise InvalidTagParserException(tag_name)

                start_block, token, end_block = self.cursor_upper.partition(
                    start_tag
                )  # noqa: E0203

                if not token:
                    logger.warning(f"Start token not found for tag: '{tag_name}'.")
                    raise StartTokenNotFoundParserException(tag_name)

                offset = len(start_block + token)

                if end_tag:
                    start_block, token, end_block = end_block.partition(end_tag)

                    if not token:
                        logger.warning(f"End token not found for tag: '{tag_name}'.")
                        raise EndTokenNotFoundParserException(tag_name)

                    self.field_upper = start_block
                    self.cursor_upper = end_block
                else:
                    self.field_upper, token, self.cursor_upper = end_block.partition(start_tag)
                    self.cursor_upper = token + self.cursor_upper

                self.field = self.cursor[offset:][: len(self.field_upper)]
                self.cursor = self.cursor[-len(self.cursor_upper) :]

                return self

        logger.warning(f"Tag '{tag_name}' not found during parsing.")
        raise TagNotFoundParserException(tag_name)
